### PR TITLE
Toolbar: Use Ariakit instead of Reakit

### PR DESCRIPTION
**Pull request by diegohaz**
## What?

This PR updates the Toolbar component from the `@wordpress/components` package to use Ariakit instead of Reakit. [Ariakit](https://ariakit.org) is the successor of [Reakit](https://reakit.io).

## Why?

Ariakit is already superior to Reakit in all aspects. It is faster, offers more features, improvements, and has fewer bugs. In fact, dozens of bugs have been reported in the past year, and all of them could be resolved in userland before a fix was implemented in the library. See [closed bugs](https://github.com/ariakit/ariakit/issues?q=is%3Aissue+label%3A%22bug%22+is%3Aclosed+) and notice that all of them have the "has workaround" label.

Ariakit provides full support for the latest versions of React (17 and 18). It is still released as v0.x.x because we are still working on the website and preparing for an official release.

Breaking changes are released as v0.x.0, following semver. Additionally, most of these changes are preceded by soft deprecations (without warnings), followed by deprecation warnings before the actual breaking change.

## How?

I updated the Toolbar component to use the new API (mostly `state` → `store`). I've also updated the `ToolbarItem` prop types.

## Testing Instructions

The Toolbar component is mainly used in the `Block tools` toolbar (the block one) and the `Document tools` toolbar (the top one) on the editor. There should be no change in how they work.

Those components are covered by the following e2e tests:

- `navigable-toolbar.spec.js`
- `shortcut-focus-toolbar.spec.js`
- `toolbar-roving-tabindex.spec.js`

All tests should pass.

### Testing Instructions for Keyboard

The `shortcut-focus-toolbar.spec.js` and `toolbar-roving-tabindex.spec.js` tests cover most of the keyboard interactions on the toolbar.

Make sure you can navigate to the block toolbar using either <kbd>Shift</kbd><kbd>Tab</kbd> on the block or <kbd>Alt</kbd><kbd>F10</kbd>. Once in the toolbar, you should be able to navigate through the items using arrow keys and interact with buttons by pressing <kbd>Enter</kbd> and <kbd>Space</kbd>. Dropdowns should open on the <kbd>ArrowDown</kbd> key and close with <kbd>Esc</kbd>, returning the focus to the dropdown button in the toolbar.

Interactions should also remain unchanged when the `Top toolbar` setting is enabled (in the `Options` menu located at the end of the `Editor top bar` region). **Note**: As a sighted keyboard user, I noticed the behavior here is quite confusing. The visual position of the elements in the top toolbar (that now has both the block tools and document tools) doesn't seem to match the tab order. But this behavior is also present on trunk, so I guess it's out of the scope of this PR.

### CODE REVIEW

**Comment by alexstine** [comment]
@diegohaz Please ping me when this is ready for accessibility testing. I doubt there will be any regressions but just to be sure. 👍 Thanks.

**Comment by diegohaz** [comment]
Thank you, @alexstine! It's ready for review now.

**Comment by diegohaz** [comment]
> | Filename | Size | Change |
> | :--- | :---: | :---: |
> | `build/components/index.min.js` | 240 kB | +9.35 kB (+4%) |

This increase in size should be mitigated when Reakit is fully replaced. I'll work on migrating the other parts after this PR is merged.

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (tyxla) [inline packages/components/src/toolbar/toolbar-item/types.ts:23]: Seems like we document this as a render prop, but we're also accepting a `ReactNode` - should we reflect that in the inline docs above?

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (tyxla) [inline packages/components/src/toolbar/toolbar-item/types.ts:16]: I think we've typed the `as` prop like this before:

```suggestion
	as?: ElementType | keyof JSX.IntrinsicElements;
```

**Review by tyxla — APPROVED** [review]
The code looks great, and it appears to be testing well! Exciting to see us migrating to `ariakit`. Nice work - thanks @diegohaz! 🚀 

Just in case I'm missing something, would be nice to get a 👍 from @ciampo or @mirka.

**Review by alexstine — APPROVED** [review]
@diegohaz This is testing well. 👍

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (diegohaz) [reply packages/components/src/toolbar/toolbar-item/types.ts:23]: Thanks for the review @tyxla! I've updated the docs.

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (diegohaz) [reply packages/components/src/toolbar/toolbar-item/types.ts:16]: `ElementType` already includes `keyof JSX.IntrinsicElements`.

**Comment by diegohaz** [comment]
Thank you so much for the reviews. I'll wait for @ciampo or @mirka's approval before merging this.

While I understand the team is currently experimenting with Radix UI, migrating the components that currently rely on Reakit to Ariakit, now that Ariakit has achieved stability, will leave the project in a much better state.

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (ciampo) [inline packages/components/src/toolbar/toolbar-item/types.ts]: We've been using `type` for components props in the package, should we continue using `type` for consistency?

**Review by ciampo — COMMENTED** [review]
Thank you for working on this refactor, Diego ❤️ 

Code changes LGTM, will test changes in the editor soon.

> While I understand the team is currently experimenting with Radix UI, migrating the components that currently rely on Reakit to Ariakit, now that Ariakit has achieved stability, will leave the project in a much better state.

We are indeed experimenting with using Radix UI with some of our components, although that doesn't mean that we won't keep using Reakit/Ariakit, especially for its Toolbar-related components.

Are you planning on migrating the remaining `reakit` instances in the repo to `ariakit`, so that we can completely remove the `reakit` dependency from the repo?

**Review by ciampo — COMMENTED** [review]
Thank you for working on this refactor, Diego ❤️ 

Code changes LGTM, will test changes in the editor soon.

> While I understand the team is currently experimenting with Radix UI, migrating the components that currently rely on Reakit to Ariakit, now that Ariakit has achieved stability, will leave the project in a much better state.

We are indeed experimenting with using Radix UI with some of our components, although that doesn't mean that we won't keep using Reakit/Ariakit, especially for its Toolbar-related components.

Are you planning on migrating the remaining `reakit` instances in the repo to `ariakit`, so that we can completely remove the `reakit` dependency from the repo?

**Review by ciampo — COMMENTED** [review]
Thank you for working on this refactor, Diego ❤️ 

Code changes LGTM, will test changes in the editor soon.

> While I understand the team is currently experimenting with Radix UI, migrating the components that currently rely on Reakit to Ariakit, now that Ariakit has achieved stability, will leave the project in a much better state.

We are indeed experimenting with using Radix UI with some of our components, although that doesn't mean that we won't keep using Reakit/Ariakit, especially for its Toolbar-related components.

Are you planning on migrating the remaining `reakit` instances in the repo to `ariakit`, so that we can completely remove the `reakit` dependency from the repo?

**Review by ciampo — COMMENTED** [review]
Thank you for working on this refactor, Diego ❤️ 

Code changes LGTM, will test changes in the editor soon.

> While I understand the team is currently experimenting with Radix UI, migrating the components that currently rely on Reakit to Ariakit, now that Ariakit has achieved stability, will leave the project in a much better state.

We are indeed experimenting with using Radix UI with some of our components, although that doesn't mean that we won't keep using Reakit/Ariakit, especially for its Toolbar-related components.

Are you planning on migrating the remaining `reakit` instances in the repo to `ariakit`, so that we can completely remove the `reakit` dependency from the repo?

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (diegohaz) [reply packages/components/src/toolbar/toolbar-item/types.ts]: I updated it to use `type` to keep it consistent, but I recommend changing the convention to use `interface` when extending other types.

Intersection types may silently introduce incompatible types that will only show an error at usage. Also, interfaces merge JSDoc better: [TS playground with more details](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgCQCoFkAyBBGMrABGArjAKYDOANHAEpkCGAxjAHIQAmZcAvnAGZQIIOACIojFqIDcAKFkB6BXBgBPMNwDuwGAAs4wAHbkoFMi2ARDFOCAaq4EQhSbEJcEgHMKstRrgAysAANmTGcAC8KBg4eAQk5BQAPGhYACLAAG4AoqEgYTAAfHAAZIiycHBMuiEcEoYA-ABcdJJsnNwAPnAAFD0AlJHFCDz9sjzySiq6DPAM7lbBDgDWhhCaho6bxBQMnmSyTFYU8NW19S1BocYA2qJnwXVhogC6kb2DEcM8corKRiZ+MwtDp9GQAB7kQwcGzaYLBOAUXTrabcIxHcCzIghHQOAieXRzTT2WQAshQIFMbjZcFgYLAJg6OAQqEw6JYXD4IikSgpGIZHJ5ArDCpVGqPerNVrMdpcODdPqfb79OQTP6IiBwTTcJgMTb8YDggwwUnGcmU7gAYWgEhYzMhYTZAHkQDo+Ry4tzEu7MALcmR8sZCjR7uKnoZRCLKg9w1L6DL2HKFQMhohRqrJv8zaZzDBLJs-JQ4AxghRNRwrAByeAzaGhOAAKQCaQgTDgmXJDm18N86m42HeCFFCgAVCPRXAR3A0pQmAQwHmrBOpwABLhA4jBeD4YgHSqThSisBCMAtQgQCChPUZ9XAfharR67eaiAdqAEOU6Gj8aD2hjgUJe38AAhd4BzKIdKlHVd1wYTd4CBUtuBHQ9KmPCBTw8C8r0MG8pj0bgmxbNs4QRBgwA0eY4A4Yg6QZWYyA4XpkTfRxWOAu50LAV4PDIYJ1jGI5rHgQgWg40QuJ4qJELMX4pjJClgTgWsOFCGwmRAYhql4vByVNQElMte1WRsAdILgaC4DXMgNy3AQSzMA8jxPM9sMYXDxkzZTX3JNjfMtTiTx4wg+PWQ5jlOFoAokoK3iiHcyGkIA).

**Comment by diegohaz** [comment]
> Are you planning on migrating the remaining `reakit` instances in the repo to `ariakit`, so that we can completely remove the `reakit` dependency from the repo?

Yes, that's the plan.

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (ciampo) [reply packages/components/src/toolbar/toolbar-item/types.ts]: Thank you. That is a great point and something we may want to look at improving (cc @mirka )

**Review by ciampo — APPROVED** [review]
LGTM 🚀

File: packages/components/src/toolbar/toolbar-item/types.ts
Comment (mirka) [reply packages/components/src/toolbar/toolbar-item/types.ts]: Yes, although there is a certain simplicity to sticking with `type` everywhere, I'm open to the "use `interface` unless something requires a `type`" guideline, suggested in the official docs and also pretty much at [Google](https://google.github.io/styleguide/tsguide.html#interfaces-vs-type-aliases).

No need to overhaul anything in bulk, but maybe we can start by preferring `interface` in our own patches and not suggesting changes when a contributor uses `interface`?

Comment (ciampo) [reply packages/components/src/toolbar/toolbar-item/types.ts]: > No need to overhaul anything in bulk, but maybe we can start by preferring interface in our own patches and not suggesting changes when a contributor uses interface?

Sounds good.

The reason why I suggested `type` is because my personal preference is to fallback to prefer consistency when we don't have a clear guideline on the matter.

But I agree with your point! We can definitely agree to go for your approach about `interface` from now on.

Comment (mirka) [reply packages/components/src/toolbar/toolbar-item/types.ts]: > fallback to prefer consistency when we don't have a clear guideline on the matter

Me too. Let's try this out low key, and we can add it to the CONTRIBUTING guidelines when we feel a bit more confident about it 😄

**Pull request merged**
